src/plugins/irc/irc-server.c

```diff
@@ -4313,75 +4283,188 @@ irc_server_check_certificate_fingerprint (struct t_irc_server *server,
-
-    for (i = 0; i < IRC_FINGERPRINT_NUM_ALGOS; i++)
-    {
-        fingerprint_server[i] = NULL;
-    }
+    size_t j, size_bytes, pkbuf_len;
+    gnutls_pubkey_t pk;
 
     /* split good_fingerprints */
     fingerprints = weechat_string_split (good_fingerprints, ",", NULL,
                                          WEECHAT_STRING_SPLIT_STRIP_LEFT
                                          | WEECHAT_STRING_SPLIT_STRIP_RIGHT
                                          | WEECHAT_STRING_SPLIT_COLLAPSE_SEPS,
                                          0, NULL);
+
     if (!fingerprints)
         return 0;
 
     rc = 0;
+    pkbuf = NULL;
+    pkbuf_len = 0;
 
-    for (i = 0; fingerprints[i]; i++)
+    for (algo = 0; algo < IRC_FINGERPRINT_NUM_ALGOS; algo++)
     {
-        size_bits = strlen (fingerprints[i]) * 4;
-        size_bytes = size_bits / 8;
-
-        algo = irc_server_fingerprint_search_algo_with_size (size_bits);
-        if (algo < 0)
-            continue;
+        certfpstr[algo] = NULL;
+        keyfpstr[algo] = NULL;
+    }
 
-        if (!fingerprint_server[algo])
+    /* attempt to extract and fingerprint the certificate's public key */
+    if (gnutls_pubkey_init (&pk) != GNUTLS_E_SUCCESS)
+    {
+        weechat_printf (
+            server->buffer,
+            _("%sgnutls: failed to initialize public key object"),
+            weechat_prefix ("error"));
+    }
+    else
+    {
+        if (gnutls_pubkey_import_x509 (pk, certificate, 0)
+            != GNUTLS_E_SUCCESS)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to import certificate public key"),
+                weechat_prefix ("error"));
+        }
+        else if (gnutls_pubkey_export (pk, GNUTLS_X509_FMT_DER, NULL,
+                     &pkbuf_len) != GNUTLS_E_SHORT_MEMORY_BUFFER)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to export certificate public key"),
+                weechat_prefix ("error"));
+        }
+        else if (!pkbuf_len)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to export certificate public key"),
+                weechat_prefix ("error"));
+        }
+        else if (!(pkbuf = malloc (pkbuf_len)))
+        {
+            weechat_printf (
+                server->buffer,
+                _("%s%s: not enough memory (%s)"),
+                weechat_prefix ("error"), IRC_PLUGIN_NAME,
+                "fingerprint");
+        }
+        else if (gnutls_pubkey_export (pk, GNUTLS_X509_FMT_DER, pkbuf,
+                     &pkbuf_len) != GNUTLS_E_SUCCESS)
+        {
+            weechat_printf (
+                server->buffer,
+                _("%sgnutls: failed to export certificate public key"),
+                weechat_prefix ("error"));
+        }
+        else for (algo = 0; algo < IRC_FINGERPRINT_NUM_ALGOS; algo++)
```

Comment:
I think we don't use this `else for` style anywhere.

Reply:
People use `else if` all the time, but that isn't actually a syntactical construct.

This:

```C
if (foo)
    // ...
else if (bar)
    // ...
else if (baz)
    // ...
else
    // ...
```

Is just a lesser-indentation way of writing:

```C
if (foo)
    // ...
else
    if (bar)
        // ...
    else
        if (baz)
            // ...
        else
            // ...
```

They are the same. I don't see any reason not to do the same for `else for`; it just saves a layer of unnecessary indentation.

Reply:
Not seeing that in the style guide which the rest of the code appears to follow.

https://weechat.org/files/doc/stable/weechat_dev.en.html#coding_c_style

Reply:
Sure-sure, but if you search through the codebase then there are at least 29 instances of a for loop directly in the else branch and that uses a properly curly-brace-separated block instead of `else for`, which we never use. It's completely unnecessary to introduce a new obscure and mixed style to our codebase.

Reply:
Then I suggest the coding style be updated to either allow or forbid it, and I will refactor appropriately.

Reply:
Resolved in 73f890030199834be2ad7be2eaaa25c87132bab2